File: 2015/ruby/day13.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end

data = File.read('day13.txt')

sample = <<~TXT
Alice would gain 54 happiness units by sitting next to Bob.
Alice would lose 79 happiness units by sitting next to Carol.
Alice would lose 2 happiness units by sitting next to David.
Bob would gain 83 happiness units by sitting next to Alice.
Bob would lose 7 happiness units by sitting next to Carol.
Bob would lose 63 happiness units by sitting next to David.
Carol would lose 62 happiness units by sitting next to Alice.
Carol would gain 60 happiness units by sitting next to Bob.
Carol would gain 55 happiness units by sitting next to David.
David would gain 46 happiness units by sitting next to Alice.
David would lose 7 happiness units by sitting next to Bob.
David would gain 41 happiness units by sitting next to Carol.
TXT

def parse_line(line)
  tokens = line.split(' ')
  a = tokens[0]
  b = tokens[-1].gsub('.', '')
  value = tokens[3].to_i * (tokens[2] == 'gain' ? 1 : -1)
  [a, b, value]
end
def combos(input)
  rv = Hash.new do |h,k|
    h[k] = Hash.new(0)
  end
  input.lines.map(&:strip).each do |line|
    a, b, value = parse_line(line)
    rv[a][b] = value
  end
  rv
end

def total_change(input, include_me=false)
  combos = combos(input)
  names = combos.keys
  names += ['***ME***'] if include_me
  names.permutation.map do |order|
    i = -1
    sum = 0
    while i < order.length - 1;
      v = combos[order[i]][order[i+1]] + combos[order[i+1]][order[i]]
      sum += v
      i += 1
    end
    sum
  end.max
end

assert_equal(['Alice', 'Bob', 54], parse_line('Alice would gain 54 happiness units by sitting next to Bob.'))
assert_equal(['Alice', 'Bob', -2], parse_line('Alice would lose 2 happiness units by sitting next to Bob.'))
assert_equal(330, total_change(sample))

puts "Part 1: #{total_change(data)}"
puts "Part 2: #{total_change(data, true)}"

File: 2015/go/day13/day13.go
package day13

import (
	"fmt"
	"strings"
)

type Instruction struct {
	To     string
	From   string
	Amount int
}

func ParseLine(line string) Instruction {
	var a, b, gl string
	var value int
	fmt.Sscanf(line, "%s would %s %d happiness units by sitting next to %s", &a, &gl, &value, &b)
	if gl == "lose" {
		value *= -1
	}
	b = strings.TrimSuffix(b, ".")
	return Instruction{a, b, value}
}

func Combos(Input []string) map[string]map[string]int {
	rv := make(map[string]map[string]int)
	for _, line := range Input {
		i := ParseLine(line)
		if _, ok := rv[i.From]; !ok {
			rv[i.From] = make(map[string]int)
		}
		rv[i.From][i.To] = i.Amount
	}

	return rv
}

func TotalChange(Input []string, IncludeMe bool) int {
	combos := Combos(Input)
	names := make([]string, 0)
	for k := range combos {
		names = append(names, k)
	}
	if IncludeMe {
		names = append(names, "***ME***")
	}
	max := 0
	for _, order := range Permutations(names) {

		i := 0
		sum := 0
		for i < len(order) {
			other := i + 1
			if other == len(order) {
				other = 0
			}

			v := combos[order[i]][order[other]] + combos[order[other]][order[i]]
			sum += v
			i += 1
		}
		if sum > max {
			max = sum
		}
	}
	return max
}

func Permutations(arr []string) [][]string {
	var helper func([]string, int)
	res := [][]string{}

	helper = func(arr []string, n int) {
		if n == 1 {
			tmp := make([]string, len(arr))
			copy(tmp, arr)
			res = append(res, tmp)
		} else {
			for i := 0; i < n; i++ {
				helper(arr, n-1)
				if n%2 == 1 {
					tmp := arr[i]
					arr[i] = arr[n-1]
					arr[n-1] = tmp
				} else {
					tmp := arr[0]
					arr[0] = arr[n-1]
					arr[n-1] = tmp
				}
			}
		}
	}
	helper(arr, len(arr))
	return res
}
